File: aihw-ppci-compiler/atalla_tests/kernels/relu.c
/*
 * ReLU kernel: max(0, x) element-wise on a tile loaded from DRAM.
 *
 * Config at ADDR_TABLE (0x3C):
 *   [0] IN_GMEM    [4] OUT_GMEM
 *
 * Fixed tile: 4 rows x 32 cols (128 bf16 elements).
 * Loads tile from DRAM into scratchpad, zeroes negatives, stores back.
 */

#define CFG_BASE  0x3C
#define WIDTH_M1  31
#define ROWS      4
#define ROWS_M1   3
#define ALL_MASK  0xFFFFFFFF

int main() {
    int cfg = CFG_BASE;
    int IN_GMEM;
    int OUT_GMEM;
    asm("lw_s %0, 0(%1)" : "=r"(IN_GMEM)  : "r"(cfg));
    asm("lw_s %0, 4(%1)" : "=r"(OUT_GMEM) : "r"(cfg));

    int sp = 0;
    int all_mask = ALL_MASK;
    int ncols = 1;
    int sdma_ctl;
    asm("li_s %0, 133169183" : "=r"(sdma_ctl));

    /* load input tile from DRAM to scratchpad */
    asm("scpad_ld %0, %1, %2" : : "r"(sp), "r"(IN_GMEM), "r"(sdma_ctl));

    /* create zero vector: load any row then multiply by 0 */
    vec zero_vec;
    asm("vreg_ld %0, %1, %2, 31, 0" : "=v"(zero_vec) : "r"(0), "r"(ncols));
    zero_vec = vec_op_masked("*", zero_vec, 0.0, all_mask);

    int row = 0;
    while (row < ROWS) {
        vec v;
        asm("vreg_ld %0, %1, %2, 31, 0" : "=v"(v) : "r"(row), "r"(ncols));

        /* build mask of negative elements, then zero them out */
        int m_neg = make_mask("<", v, zero_vec, all_mask);
        vec result = vec_op_masked("*", v, 0.0, m_neg);

        asm("vreg_st %0, %1, %2, 31, 0" : : "v"(result), "r"(row), "r"(ncols));
        row = row + 1;
    }

    /* store result tile back to DRAM */
    asm("scpad_st %0, %1, %2" : : "r"(sp), "r"(OUT_GMEM), "r"(sdma_ctl));

    asm("halt");
    return 0;
}

File: atalla-graph/kernels/relu.py
"""AtallaC ReLU kernel: element-wise max(x, 0) via masked zero."""

import math
from kernels.common import ADDR_TABLE, TILE, sdma_ctl_expr


def relu_c(total: int, width: int) -> str:
    """Generate AtallaC for ReLU."""
    rows = math.ceil(total / width)
    w_m1 = width - 1
    sp_rows = min(rows, TILE)
    tile_count = math.ceil(rows / sp_rows)
    tile_bytes = sp_rows * width * 2
    sdma_s = sdma_ctl_expr("sdma_ctl", 0, sp_rows, width, width)

    return (
        "int main() {\n"
        f"    int cfg = {ADDR_TABLE};\n"
        "    int IN_GMEM;\n"
        "    int OUT_GMEM;\n"
        '    asm("lw_s %0, 0(%1)" : "=r"(IN_GMEM)  : "r"(cfg));\n'
        '    asm("lw_s %0, 4(%1)" : "=r"(OUT_GMEM) : "r"(cfg));\n'
        "\n"
        "    int sp = 0;\n"
        "    int all_mask = -1;\n"
        "    int ncols = 1;\n"
        f"{sdma_s}"
        "\n"
        f"    vec zero_vec = vector_load(sp, 0, {w_m1}, 0);\n"
        '    zero_vec = vec_op_masked("*", zero_vec, 0.0, all_mask);\n'
        "\n"
        "    int tile = 0;\n"
        f"    while (tile < {tile_count}) {{\n"
        "        scpad_load(sp, IN_GMEM, sdma_ctl);\n"
        "\n"
        "        int row = 0;\n"
        f"        while (row < {sp_rows}) {{\n"
        f"            vec v = vector_load(sp, row, {w_m1}, 0);\n"
        "\n"
        '            int m_neg = make_mask("<", v, zero_vec, all_mask);\n'
        '            vec result = vec_op_masked("*", v, 0.0, m_neg);\n'
        "\n"
        f"            vector_store(result, sp, row, {w_m1}, 0);\n"
        "            row = row + 1;\n"
        "        }\n"
        "\n"
        "        scpad_store(sp, OUT_GMEM, sdma_ctl);\n"
        f"        IN_GMEM = IN_GMEM + {tile_bytes};\n"
        f"        OUT_GMEM = OUT_GMEM + {tile_bytes};\n"
        "        tile = tile + 1;\n"
        "    }\n"
        "\n"
        '    asm("halt");\n'
        "    return 0;\n"
        "}\n"
    )
